src/metabase/api/collection.clj

```diff
@@ -45,17 +45,54 @@
 
 (declare root-collection)
 
+(defn- remove-other-users-personal-collections
+  "Remove from the provided collections any that are personal and belong to another user.
+   Will work in a single pass if collections are already ordered such that root (location = /)
+   collections appear before children, but will catch any mis-ordered entries and
+   do a final pass on those if needed."
+  ([user-id collections]
+   (loop [[{:keys [id location personal_owner_id] :as c} & r] collections
+          keep-roots     #{user-id}
+          skip-roots     #{}
+          res            []
+          indeterminates []]
+     (if c
```

Comment:
there are lots of conditions here and it takes time to follow this, maybe you could add some comments explaining what each branch does?

Reply:
Done